### PR TITLE
Login: Jetpack setup error handling

### PR DESCRIPTION
**Pull request by itsmeichigo**
Part of #8075 

### Description

This PR adds error handling for the native Jetpack setup flow. There are 3 cases handled here:

**1. Jetpack connection authorized with a WP.com account other than the logged-in one.**
When the user taps the "Log in with another account" link on the Jetpack connection web page, they can log in with a different account other than the logged-in one. I handled this edge case by:
- Updated `SiteCredentialLoginViewModel` to return the site XMLRPC URL when the login completes.
- Updated the `LoginJetpackSetupViewModel` to return the Jetpack-connected email when the setup completes.
- Updated the coordinator to navigate to WP.com account verification if the Jetpack-connected email is different from the logged-in email.

**2. Error handling for setup steps.**
- Updated `LoginJetpackSetupViewModel` with information about the setup failure, error messages and codes, and visibility for error state views.
- `checkJetpackConnection` is handled with an auto-retry mechanism that allows a maximum of 2 retries. This is to ensure that Jetpack is synced properly for the site and returns the correct connection status.
- Updated `LoginJetpackSetupView` to show appropriate contents for error states.

**3. Handling interrupted Jetpack connection.**
When the user dismisses the web view for the Jetpack connection, we should show a modal indicating that the connection step has been interrupted. A new SwiftUI view has been added for this: `LoginJetpackSetupInterruptedView`.

Due to the size of the PR, unit tests will be added in a separate PR.

### Testing instructions

Prerequisite: make sure that you have a test site without Jetpack.

**1. Jetpack connection authorized with a WP.com account other than the logged-in one.**
- Log out of the app or skip login onboarding if needed.
- On the prologue screen, select "Log In" or "Continue with WordPress.com" based on the A/B test variant you get.
- Log in with your WordPress.com account.
- On the store picker, select Add a Store > Connect an existing site.
- Enter the address of your test store and tap Continue.
- On the Jetpack error screen, select Install Jetpack or Connect Jetpack.
- When the site credential login screen is presented, enter the correct credentials for your store.
- On the Jetpack setup screen, when the setup reaches the connection step, notice that a web view is presented for approving Jetpack connection.
- Tap the "Log in with another account" on the web page and proceed to log in with a WP.com other than your logged-in account.
- Tap on the Authorize button on the web page, notice that the web view is then dismissed, and the Jetpack setup screens should then show that all steps are completed.
- Tap on Go to Store, notice that you are navigated to a new screen for verifying the password for the account you just entered on the web page. 
- After the login succeeds, notice that you are navigated to the home screen.

**2. Error handling for an account without permission.**
- Log out of the app or skip login onboarding if needed.
- On the prologue screen, select "Log In" or "Continue with WordPress.com" based on the A/B test variant you get.
- Log in with your WordPress.com account.
- On the store picker, select Add a Store > Connect an existing site.
- Enter the address of your test store and tap Continue.
- On the Jetpack error screen, select Install Jetpack or Connect Jetpack.
- When the site credential login screen is presented, enter the correct credentials for an account **whose role is neither shop manager or administrator.**
- On the Jetpack setup screen, notice that the error state is then be displayed with a message regarding a permission issue. The retry CTA should not be available for this error. The Cancel and Get support CTAs should work properly.

**3. Error handling for issues other than permission**
- Log out of the app or skip login onboarding if needed.
- On the prologue screen, select "Log In" or "Continue with WordPress.com" based on the A/B test variant you get.
- Log in with your WordPress.com account.
- On the store picker, select Add a Store > Connect an existing site.
- Enter the address of your test store and tap Continue.
- On the Jetpack error screen, select Install Jetpack or Connect Jetpack.
- When the site credential login screen is presented, enter the correct credentials for a shop manager or administrator account.
- When the Jetpack setup screen appears, turn off the internet connection or intercept the response with a 5xx error. Notice that the error state will then be displayed with appropriate messages. The retry CTA should be available and work properly.

**4. Error handling for interrupted Jetpack connection**
- Log out of the app or skip login onboarding if needed.
- On the prologue screen, select "Log In" or "Continue with WordPress.com" based on the A/B test variant you get.
- Log in with your WordPress.com account.
- On the store picker, select Add a Store > Connect an existing site.
- Enter the address of your test store and tap Continue.
- On the Jetpack error screen, select Install Jetpack or Connect Jetpack.
- When the site credential login screen is presented, enter the correct credentials for a shop manager or administrator account.
- On the Jetpack setup screen, when the setup reaches the connection step, notice that a web view is presented for approving the Jetpack connection.
- Tap Cancel to dismiss the web view, notice that a new view is presented regarding the interrupted Jetpack connection.
- Tap Continue Connection, notice that the error screen is dismissed, and the web view is displayed again.
- Repeat tapping Cancel, this time tap Help, notice that the support screen is displayed.
- Tap the Cancel Installation button, notice that the installation flow is dismissed.

### Screenshots

**1. Jetpack connection authorized with a WP.com account other than the logged-in one.**

https://user-images.githubusercontent.com/5533851/203531637-52ce0fc7-63da-429e-ba7b-3989bd5ca038.mp4

**2. Error handling for an account without permission.**

https://user-images.githubusercontent.com/5533851/203531829-3b708d4a-9c1f-416a-8010-da30d76f55e8.mp4

**3. Error handling for issues other than permission**

https://user-images.githubusercontent.com/5533851/203531940-2e512b8a-e6be-41a8-8786-e3980c134a94.mp4

**4. Error handling for interrupted Jetpack connection**

https://user-images.githubusercontent.com/5533851/203532077-50ccd77e-5707-4b31-a90e-efb19e00cd96.mp4

---
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/LoginJetpackSetupInterruptedView.swift [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+///
+struct LoginJetpackSetupInterruptedView: View {
+    let onSupport: () -> Void
+    let onContinue: () -> Void
+    let onCancellation: () -> Void
+
+    var body: some View {
+        GeometryReader { geometry in
+            ScrollView {
+                VStack {
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/LoginJetpackSetupInterruptedView.swift]: Nit: Is it possible to use `ScrollableVStack` here?

File: WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/LoginJetpackSetupView.swift [hunk 2]
```diff
@@ -3,14 +3,27 @@ import SwiftUI
...
+
         rootView.webViewPresentationHandler = { [weak self] in
             self?.presentJetpackConnectionWebView()
         }
+
+        rootView.supportHandler = { [weak self] in
+            guard let self else { return }
+            // dismiss any presented view if possible
+            self.presentedViewController?.dismiss(animated: true)
+            authentication.presentSupport(from: self, screen: .jetpackRequired)
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/LoginJetpackSetupView.swift]: Super nit: Consider moving this to a private method just like other handlers.

File: WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/LoginJetpackSetupView.swift [hunk 3]
```diff
@@ -100,7 +124,7 @@ struct LoginJetpackSetupView: View {
                     ActivityIndicator(isAnimating: .constant(true), style: .medium)
                     Spacer()
                 }
-                .renderedIf(viewModel.currentSetupStep == nil)
+                .renderedIf(viewModel.shouldShowInitialLoadingIndicator)
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/LoginJetpackSetupView.swift:137]: Nice 👍

File: WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/LoginJetpackSetupView.swift [hunk 4]
```diff
@@ -176,6 +271,8 @@ private extension LoginJetpackSetupView {
         static let stepItemHorizontalSpacing: CGFloat = 24
         static let stepItemsVerticalSpacing: CGFloat = 20
         static let stepImageSize: CGFloat = 24
+        static let supportImageSize: CGFloat = 18
+        static let errorContentSpacing: CGFloat = 16
     }
 }
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/LoginJetpackSetupView.swift:296]: Super nit: Is it possible to add a view to `LoginJetpackSetupView_Previews` to show the error scenario?

File: WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/LoginJetpackSetupView.swift [hunk 5]
```diff
@@ -153,21 +210,59 @@ struct LoginJetpackSetupView: View {
...
         .onChange(of: viewModel.shouldPresentWebView) { shouldPresent in
             if shouldPresent {
                 webViewPresentationHandler()
             }
         }
+        .fullScreenCover(isPresented: $viewModel.jetpackConnectionInterrupted) {
+            LoginJetpackSetupInterruptedView(onSupport: supportHandler, onContinue: {
+                viewModel.jetpackConnectionInterrupted = false
+                // delay for 300ms for the dismissal of the interrupted screen to complete.
+                DispatchQueue.main.asyncAfter(deadline: .now() + 0.3) {
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/LoginJetpackSetupView.swift]: Should we move `0.3` to a `Constants` enum?

File: WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/LoginJetpackSetupView.swift [hunk 6]
```diff
@@ -139,8 +163,41 @@ struct LoginJetpackSetupView: View {
...
+                        } icon: {
+                            Image(systemName: "questionmark.circle")
+                                .resizable()
+                                .frame(width: Constants.supportImageSize * scale, height: Constants.supportImageSize * scale)
+                        }
+                    }
+                    .buttonStyle(.plain)
+                    .foregroundColor(Color(uiColor: .withColorStudio(.blue, shade: .shade50)))
 
+                }
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/LoginJetpackSetupView.swift]: Nit: Moving this to a private view will help with readability. Or kindly consider adding comments to label the view.

File: WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/LoginJetpackSetupView.swift [hunk 7]
```diff
@@ -153,21 +210,59 @@ struct LoginJetpackSetupView: View {
...
+                .renderedIf(viewModel.hasEncounteredPermissionError == false)
+
+                Button {
+                    // TODO: add tracks
+                    cancellationHandler()
+                } label: {
+                    Text(Localization.cancelInstallation)
+                }
+                .buttonStyle(SecondaryButtonStyle())
+            }
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/LoginJetpackSetupView.swift:249]: Nit: Same as above, moving this to a private view will help with readability. Or kindly consider adding comments.

File: WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/LoginJetpackSetupViewModel.swift [hunk 8]
```diff
@@ -8,15 +9,64 @@ final class LoginJetpackSetupViewModel: ObservableObject {
...
     @Published private(set) var currentConnectionStep: ConnectionStep = .pending
     @Published private(set) var jetpackConnectionURL: URL?
     @Published var shouldPresentWebView = false
+    @Published var jetpackConnectionInterrupted = false
+
+    /// Whether the setup failed. This will be observed by `LoginJetpackSetupView` to present error modal.
+    @Published private(set) var setupFailed: Bool = false
+    @Published private(set) var setupErrorMessage: String = ""
+    @Published private(set) var setupErrorSuggestion: String = ""
+    @Published private(set) var errorCode: Int?
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/LoginJetpackSetupViewModel.swift]: Nit: If all of this all related, would it make sense to combine these into a `Struct`?

**Review by selanthiraiyan — APPROVED** [review]
Great work, Huong. 🎉 

All the scenarios described in the PR work as expected. ✅ 

I left a few non-blocking nits and questions.

File: WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/LoginJetpackSetupView.swift (hunk 4 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/LoginJetpackSetupView.swift:296]: The errors are get-only, so I'm not sure how to mock that 😅 Please let me know if you have any suggestion.

File: WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/LoginJetpackSetupInterruptedView.swift (hunk 1 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/LoginJetpackSetupInterruptedView.swift]: Good idea 👍  Updated in 8dfed739182f9daf71e1ed0e33efe09449820ac9.

File: WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/LoginJetpackSetupView.swift (hunk 2 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/LoginJetpackSetupView.swift]: Updated in 1be991101591e3f49bf29444e654ded92a100db0.

File: WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/LoginJetpackSetupView.swift (hunk 5 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/LoginJetpackSetupView.swift]: Yup, updated in f357e38413da7fa30d298e488831904610bf2428.

File: WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/LoginJetpackSetupView.swift (hunk 6 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/LoginJetpackSetupView.swift]: Added comments in 20e00e190ce5448e54fac0ba650cae9c2ccd6cd4.

File: WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/LoginJetpackSetupView.swift (hunk 7 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/LoginJetpackSetupView.swift:249]: Also updated in 20e00e190ce5448e54fac0ba650cae9c2ccd6cd4.

File: WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/LoginJetpackSetupViewModel.swift (hunk 8 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/LoginJetpackSetupViewModel.swift]: Good point! Updated in f6037222df9e2a65eaf670a1b89720f195608946.

**Comment by itsmeichigo** [comment]
Thank you @selanthiraiyan for the review! I updated the code following your suggestions above.

I also added assets for dark mode in ab1613f2b6fd7e28327422b4d5ee6cbeead85038 - sorry for cramping this here since CI might take a long time to handle several PRs 😅 .

Screenshots for dark mode:

| Jetpack setup | Jetpack connection | Interrupted connection |
| ----- | ----- | ----- |
|  |  |  |

Updated light mode asset for interrupted connection:

**Comment by selanthiraiyan** [comment]
Thanks for the updates. This still LGTM. 🚢

**Pull request merged**